### PR TITLE
Support collate in LowCardinality, Nullable, Array and Tuple, where nested type is String

### LINKED ISSUE #15534: LowCardinality doesn't work with ORDER BY COLLATE
**Describe the bug**
You can't use COLLATE with LowCardinality(String) in it.
**How to reproduce**
Clickhouse server version  20.10.1,  20.3.10.75
```
SELECT
    toLowCardinality(arrayJoin(splitByString(', ', 'ABC, bA, Ba, a b c, abc, ba, BA'))) AS a,
    toTypeName(a)
ORDER BY a ASC COLLATE 'tr'

↓ Progress: 1.00 rows, 1.00 B (9.81 rows/s., 9.81 B/s.)
Received exception from server (version 20.10.1):
Code: 262. DB::Exception: Received from localhost:9000. DB::Exception: Collations could be specified only for String columns..

0 rows in set. Elapsed: 0.103 sec.


SELECT
    toLowCardinality(arrayJoin(splitByString(', ', 'ABC, bA, Ba, a b c, abc, ba, BA'))) AS a,
    toTypeName(a)
ORDER BY CAST(a, 'String') ASC COLLATE 'tr'

┌─a─────┬─toTypeName(toLowCardinality(arrayJoin(splitByString(', ', 'ABC, bA, Ba, a b c, abc, ba, BA'))))─┐
│ a b c │ LowCardinality(String)                                                                          │
│ abc   │ LowCardinality(String)                                                                          │
│ ABC   │ LowCardinality(String)                                                                          │
│ ba    │ LowCardinality(String)                                                                          │
│ bA    │ LowCardinality(String)                                                                          │
│ Ba    │ LowCardinality(String)                                                                          │
│ BA    │ LowCardinality(String)                                                                          │
└───────┴─────────────────────────────────────────────────────────────────────────────────────────────────┘

7 rows in set. Elapsed: 0.003 sec.
```

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add COLLATE support for Nullable, LowCardinality, Array and Tuple, where nested type is String. Also refactor the code associated with collations in ColumnString.cpp.

This closes https://github.com/ClickHouse/ClickHouse/issues/15534

### CODE REVIEW

File: src/Columns/ColumnLowCardinality.cpp [hunk 1]
```diff
@@ -295,14 +297,24 @@ void ColumnLowCardinality::compareColumn(const IColumn & rhs, size_t rhs_row_num
...
     if (limit == 0)
         limit = size();
 
     size_t unique_limit = getDictionary().size();
     Permutation unique_perm;
-    getDictionary().getNestedColumn()->getPermutation(reverse, unique_limit, nan_direction_hint, unique_perm);
+    if (collator)
+    {
+        /// Collations are supported only for ColumnString
+        const ColumnString * column_string = checkAndGetColumn<ColumnString>(getDictionary().getNestedColumn().get());
```
Comment (alexey-milovidov) [inline src/Columns/ColumnLowCardinality.cpp:310]: There is such a thing as `LowCardinality(Nullable(String))`

File: src/Columns/ColumnLowCardinality.cpp [hunk 2]
```diff
@@ -295,14 +297,24 @@ void ColumnLowCardinality::compareColumn(const IColumn & rhs, size_t rhs_row_num
...
+        if (!column_string)
+            throw Exception("Collations could be specified only for String columns or columns where nested column is String.", ErrorCodes::BAD_COLLATION);
+
+        column_string->getPermutationWithCollation(*collator, reverse, unique_limit, unique_perm);
```
Comment (alexey-milovidov) [inline src/Columns/ColumnLowCardinality.cpp:314]: Everything should be simpler if we will add this method to IColumn.

File: src/Columns/ColumnLowCardinality.cpp [hunk 3]
```diff
@@ -345,20 +358,17 @@ void ColumnLowCardinality::updatePermutation(bool reverse, size_t limit, int nan
...
     for (size_t i = 0; i < number_of_ranges; ++i)
     {
         const auto& [first, last] = equal_ranges[i];
-        if (reverse)
-            std::sort(res.begin() + first, res.begin() + last, [this, nan_direction_hint](size_t a, size_t b)
-                      {return getDictionary().compareAt(getIndexes().getUInt(a), getIndexes().getUInt(b), getDictionary(), nan_direction_hint) > 0; });
-        else
-            std::sort(res.begin() + first, res.begin() + last, [this, nan_direction_hint](size_t a, size_t b)
-                      {return getDictionary().compareAt(getIndexes().getUInt(a), getIndexes().getUInt(b), getDictionary(), nan_direction_hint) < 0; });
+        std::sort(res.begin() + first, res.begin() + last, less);
```
Comment (alexey-milovidov) [inline src/Columns/ColumnLowCardinality.cpp:366]: Maybe merge-sort is better for heavy comparisons, what do you think?

File: src/Columns/ColumnString.h [hunk 4]
```diff
@@ -42,15 +42,19 @@ class ColumnString final : public COWHelper<IColumn, ColumnString>
     /// Size of i-th element, including terminating zero.
     size_t ALWAYS_INLINE sizeAt(ssize_t i) const { return offsets[i] - offsets[i - 1]; }
 
-    template <bool positive>
-    struct less;
+    struct cmp;
```
Comment (alexey-milovidov) [inline src/Columns/ColumnString.h:45]: `Cmp` with capital letter will be more in-style.

File: src/Columns/ColumnLowCardinality.cpp (hunk 3 above)
Comment (Avogar) [reply src/Columns/ColumnLowCardinality.cpp:370]: I made some experiments with std::stable_sort with large data. The results were worse than with std::sort.

**Comment by alexey-milovidov** [comment]
`string_sort` tests has slowed down 5% and `collations` has slowed down 2%. But it is below the radars.

**Comment by alexey-milovidov** [comment]
The worst is `SELECT PageCharset FROM hits_100m_single ORDER BY PageCharset, CounterID LIMIT 10 format Null`.
It is slowed down 15%.

**Pull request merged**